File: LongestIncreasingPathinaMatrix2.cpp
#include<iostream>
#include<vector>
#include<unordered_set>
#include<algorithm>

using namespace std;

class Solution {
public:
    int longestIncreasingPath(vector<vector<int> >& matrix) {
    	
    	int m = matrix.size();
    	if(m == 0) return 0;
    	int n = matrix[0].size();
    	
    	int count = m * n, ans = 0;
    	while(count > 0) {
    		vector<pair<int, int> > remove;
    		for(int i = 0; i < m; i++) {
    			for(int j = 0; j < n; j++) {
    				if(matrix[i][j] == INT_MIN) continue;
    				bool up = (i == 0 || matrix[i][j] >= matrix[i - 1][j]);
    				bool down = (i == m - 1 || matrix[i][j] >= matrix[i + 1][j]);
    				bool left = (j == 0 || matrix[i][j] >= matrix[i][j - 1]);
    				bool right = (j == n - 1 || matrix[i][j] >= matrix[i][j + 1]);
    				
    				if(up && down && left && right) remove.push_back(make_pair(i, j));
				}
			}
			
			for(auto p : remove) {
				matrix[p.first][p.second] = INT_MIN;
				count--;
			}
			
			ans++;
		}
		
		return ans;
        
    }
};

int main() {
	
	int m, n;
	cin>>m>>n;
	
	vector<vector<int> > matrix;
	
	for(int i = 0; i < m; i++) {
		vector<int> row; 
		for(int j = 0; j < n; j++) {
			int num;
			cin>>num;
			row.push_back(num);
		}
		matrix.push_back(row);
	}
	
	Solution *solution = new Solution();
	cout<<solution->longestIncreasingPath(matrix);
	
	return 0;
	
}

File: LongestIncreasingPathinaMatrix2.py
class Solution:
    def longestIncreasingPath(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: int
        """
        m = len(matrix)
        if m == 0:
            return 0
        n = len(matrix[0])

        count = m * n
        res = 0
        while count > 0:
            remove = []
            for i in range(m):
                for j in range(n):
                    if matrix[i][j] == -2147483648:
                        continue
                    up = (i == 0 or (matrix[i][j] >= matrix[i - 1][j]))
                    down = (i == m - 1 or (matrix[i][j] >= matrix[i + 1][j]))
                    left = (j == 0 or (matrix[i][j] >= matrix[i][j - 1]))
                    right = (j == n - 1 or (matrix[i][j] >= matrix[i][j + 1]))

                    if up and down and left and right:
                        remove.append((i, j))
            for p in remove:
                matrix[p[0]][p[1]] = -2147483648
                count -= 1
            res += 1
        return res


if __name__ == '__main__':
    m = int(input())
    matrix = []
    for i in range(m):
        row = [int(num) for num in input().split()]
        matrix.append(row)

    solution = Solution()
    print(solution.longestIncreasingPath(matrix))
